zerver/views/home.py

```diff
@@ -105,24 +104,15 @@ def update_last_reminder(user_profile: Optional[UserProfile]) -> None:
         user_profile.last_reminder = None
         user_profile.save(update_fields=["last_reminder"])
 
-def sent_time_in_epoch_seconds(user_message: Optional[UserMessage]) -> Optional[float]:
-    if user_message is None:
-        return None
-    # We have USE_TZ = True, so our datetime objects are timezone-aware.
-    # Return the epoch seconds in UTC.
-    return calendar.timegm(user_message.message.date_sent.utctimetuple())
-
 def get_furthest_read_time(user_profile: Optional[UserProfile]) -> Optional[float]:
-    if user_profile is None:  # nocoverage
+    if user_profile is None:
         furthest_read_time: Optional[float] = time.time()
-    elif user_profile.pointer == -1:
-        furthest_read_time = None
     else:
-        latest_read = get_usermessage_by_message_id(user_profile, user_profile.pointer)
-        if latest_read is None:
-            # Don't completely fail if your saved pointer ID is invalid
-            logging.warning("User %s has invalid pointer %s", user_profile.id, user_profile.pointer)
-        furthest_read_time = sent_time_in_epoch_seconds(latest_read)
+        user_activity = get_latest_update_message_flag_activity(user_profile)
+        if user_activity is None:
+            furthest_read_time = None
+        else:
+            furthest_read_time = calendar.timegm(user_activity.last_visit.utctimetuple())
     return furthest_read_time
```

Comment:
This function reads way better if we just `return` at each computation; just did that refactor.